tests/storages_tests/test_cached_storage.py

```diff
@@ -0,0 +1,152 @@
+
+    trial_id = storage.create_new_trial(study_id)
+    with patch.object(base_storage, "_update_trial", return_value=True) as update_mock:
+        with patch.object(base_storage, "set_trial_value", return_value=None) as set_mock:
+            storage.set_trial_value(trial_id, 0.3)
+            assert update_mock.call_count == 1
+            assert set_mock.call_count == 0
+
+    trial_id = storage.create_new_trial(study_id)
+    with patch.object(base_storage, "_update_trial", return_value=True) as update_mock:
+        with patch.object(
+            base_storage, "set_trial_intermediate_value", return_value=None
+        ) as set_mock:
+            storage.set_trial_intermediate_value(trial_id, 3, 0.3)
+            assert update_mock.call_count == 1
+            assert set_mock.call_count == 0
+
+    trial_id = storage.create_new_trial(study_id)
+    with patch.object(base_storage, "_update_trial", return_value=True) as update_mock:
+        with patch.object(base_storage, "set_trial_system_attr", return_value=None) as set_mock:
+            storage.set_trial_system_attr(trial_id, "attrA", "foo")
+            assert update_mock.call_count == 1
+            assert set_mock.call_count == 0
+
+    trial_id = storage.create_new_trial(study_id)
+    with patch.object(base_storage, "_update_trial", return_value=True) as update_mock:
+        with patch.object(base_storage, "set_trial_user_attr", return_value=None) as set_mock:
+            storage.set_trial_user_attr(trial_id, "attrB", "bar")
+            assert update_mock.call_count == 1
+            assert set_mock.call_count == 0
+
+
+def test_cache_deletion() -> None:
+
+    """Test CachedStorage deletes a cache when trial finishes."""
+
+    base_storage = RDBStorage("sqlite:///:memory:")
+    storage = _CachedStorage(base_storage)
+    study_id = storage.create_new_study("test-study")
+
+    for state in [TrialState.COMPLETE, TrialState.PRUNED, TrialState.FAIL, TrialState.WAITING]:
+        trial_id = storage.create_new_trial(study_id)
+        with patch.object(base_storage, "_update_trial", return_value=True) as update_mock:
+            with patch.object(base_storage, "set_trial_state", return_value=True) as set_mock:
+                with patch.object(base_storage, "get_trial", return_value=True) as get_mock:
+                    storage.get_trial(trial_id)
+                    assert update_mock.call_count == 0
+                    assert set_mock.call_count == 0
+                    assert get_mock.call_count == 0
+
+                    storage.set_trial_state(trial_id, state)
+                    assert update_mock.call_count == 1
+                    assert set_mock.call_count == 0
+                    assert get_mock.call_count == 0
+
+                    storage.get_trial(trial_id)
+                    assert update_mock.call_count == 1
+                    assert set_mock.call_count == 0
+                    assert get_mock.call_count == 1
```

Comment:
I think that this nest is a bit deep. How about concatenating with-statements by commas? (c.f., https://docs.python.org/3/reference/compound_stmts.html#the-with-statement)
If it's OK, please apply similar changes to other test cases.

```suggestion
        with patch.object(base_storage, "_update_trial", return_value=True) as update_mock, \
            patch.object(base_storage, "set_trial_state", return_value=True) as set_mock, \
            patch.object(base_storage, "get_trial", return_value=True) as get_mock:

            storage.get_trial(trial_id)
            assert update_mock.call_count == 0
            assert set_mock.call_count == 0
            assert get_mock.call_count == 0

            storage.set_trial_state(trial_id, state)
            assert update_mock.call_count == 1
            assert set_mock.call_count == 0
            assert get_mock.call_count == 0

            storage.get_trial(trial_id)
            assert update_mock.call_count == 1
            assert set_mock.call_count == 0
            assert get_mock.call_count == 1
```

Reply:
Thank you for the suggestion. I'll apply the change to all test cases.

Reply:
Thank you for your update!